packages/headless/src/app/commerce-ssr-engine/common.ts

```diff
@@ -0,0 +1,97 @@
+import {Controller} from '../../controllers/controller/headless-controller';
+import {InvalidControllerDefinition} from '../../utils/errors';
+import {filterObject, mapObject} from '../../utils/utils';
+import {CoreEngine, CoreEngineNext} from '../engine';
+import {InferControllerPropsMapFromDefinitions} from '../ssr-engine/types/common';
+import {
+  ControllerDefinition,
+  ControllerDefinitionOption,
+  ControllerDefinitionsMap,
+  InferControllerFromDefinition,
+  InferControllerPropsFromDefinition,
+  InferControllersMapFromDefinition,
+  SolutionType,
+} from './types/common';
+
+function buildControllerFromDefinition<
+  TControllerDefinition extends ControllerDefinition<TEngine, Controller>,
+  TEngine extends CoreEngine | CoreEngineNext,
+>({
+  definition,
+  engine,
+  solutionType,
+  props,
+}: {
+  definition: TControllerDefinition;
+  engine: TEngine;
+  solutionType: SolutionType;
+  props?: InferControllerPropsFromDefinition<TControllerDefinition>;
+}): InferControllerFromDefinition<TControllerDefinition> {
+  return (
+    'build' in definition
+      ? definition.build(engine, solutionType)
+      : definition.buildWithProps(engine, props, solutionType)
+  ) as InferControllerFromDefinition<TControllerDefinition>;
+}
+
+export function buildControllerDefinitions<
+  TControllerDefinitionsMap extends ControllerDefinitionsMap<
+    CoreEngine | CoreEngineNext,
+    Controller
+  >,
+  TEngine extends CoreEngine | CoreEngineNext,
+  TSolutionType extends SolutionType,
+>({
+  definitionsMap,
+  engine,
+  solutionType,
+  propsMap,
+}: {
+  definitionsMap: TControllerDefinitionsMap;
+  engine: TEngine;
+  solutionType: TSolutionType;
+  propsMap: InferControllerPropsMapFromDefinitions<TControllerDefinitionsMap>;
+}): InferControllersMapFromDefinition<
+  TControllerDefinitionsMap,
+  TSolutionType
+> {
+  const controllerMap = mapObject(definitionsMap, (definition, key) => {
+    // TODO: find a way to make it dynamic
+    const searchDisabled =
+      'search' in definition &&
+      definition['search'] === false &&
+      solutionType === SolutionType['search'];
```

Comment:
How is search 'disabled' if `'search'` is in the definition & the solutionType is `'search'` ?